File: BaekJoon/이분 탐색/2110_공유기 설치.py
def main():
    n, c = map(int, input().split())
    
    wifi = []
    for i in range(n):
        wifi.append(int(input()))
    
    wifi.sort()

    row = 1
    high = wifi[n-1] - wifi[0]

    ans = 0
    while row <= high:
        mid = (high + row)//2 # 공유기 간 최대 거리
        cnt = 1
        
        prev = wifi[0]
        for i in range(1, n):
            distance = wifi[i] - prev
            if distance >= mid: # 두 집 간의 거리가 공유기 간 최대 거리보다 큰 경우
                cnt += 1
                prev = wifi[i]
        
        if cnt >= c: # 공유기 간 최대 거리를 늘려야 함
            ans = mid
            row = mid + 1
        else:
            high = mid - 1
    
    print(ans)

if __name__ == "__main__":
    main()

File: BaekJoon/이분 탐색/2110_공유기 설치.swift
let inputs = readLine()!.split(separator: " ").map { Int(String($0))! }
var houses = [Int]()

for _ in 0..<inputs[0] {
    houses.append(Int(readLine()!)!)
}

houses.sort(by: <) // 오름차순으로 정렬

// Parametric Search
var left = 0
var right = 2147483647
var ans = 2147483647

while left < right {
    let mid = (left + right) / 2

    var current = houses[0] // 첫 번째 집부터
    var cnt = 1
    for i in 1..<inputs[0] {
        if houses[i] - current >= mid {
            cnt += 1
            current = houses[i] // 기준이 되는 집을 변경
        }
    }

    if cnt >= inputs[1] { // 공유기를 더 많이 설치한 경우 -> 두 집 사이의 거리를 늘려야 함
        left = mid + 1
    } else { // 공유기를 더 조금 설치한 경우 -> 두 집 사이의 거리를 줄여야 함
        right = mid
    }
}

print(left - 1)